rpcserverhandlers_test.go

```diff
@@ -3289,68 +3321,65 @@ func TestHandleSubmitBlock(t *testing.T) {
+			syncManager := defaultMockSyncManager()
+			syncManager.submitBlockErr = errors.New("block rejected")
+			return syncManager
+		}(),
 		result: "rejected: block rejected",
 	}})
 }
 
 func testRPCServerHandler(t *testing.T, tests []rpcTest) {
 	t.Helper()
 
-	for _, test := range tests {
-		cfg = defaultCfg
-		if test.mockCfg != nil {
-			cfg = test.mockCfg
-		}
+	setCfgOnce.Do(func() {
+		cfg = defaultCfg()
+	})
 
-		// Create a default rpcserverConfig and override any configurations that are
-		// provided by the test.
-		rpcserverConfig := rpcserverConfig{
-			ChainParams:  defaultChainParams,
-			Chain:        defaultRPCChain,
-			AddrManager:  defaultAddrManager,
-			SyncMgr:      defaultSyncManager,
-			ConnMgr:      defaultConnManager,
-			Clock:        defaultClock,
-			TimeSource:   blockchain.NewMedianTime(),
-			Services:     wire.SFNodeNetwork | wire.SFNodeCF,
-			SubsidyCache: standalone.NewSubsidyCache(defaultChainParams),
-		}
-		if test.mockChainParams != nil {
-			rpcserverConfig.ChainParams = test.mockChainParams
-		}
-		if test.mockChain != nil {
-			rpcserverConfig.Chain = test.mockChain
-		}
-		if test.mockAddrManager != nil {
-			rpcserverConfig.AddrManager = test.mockAddrManager
-		}
-		if test.mockSyncManager != nil {
-			rpcserverConfig.SyncMgr = test.mockSyncManager
-		}
-		if test.mockConnManager != nil {
-			rpcserverConfig.ConnMgr = test.mockConnManager
-		}
-		if test.mockClock != nil {
-			rpcserverConfig.Clock = test.mockClock
-		}
+	for _, test := range tests {
+		test := test // capture range variable
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Create a default rpcserverConfig and override any configurations that are
+			// provided by the test.
+			chainParams := defaultChainParams
+			if test.mockChainParams != nil {
+				chainParams = test.mockChainParams
+			}
+			rpcserverConfig := defaultMockConfig(chainParams)
+			if test.mockChain != nil {
+				rpcserverConfig.Chain = test.mockChain
+			}
+			if test.mockAddrManager != nil {
+				rpcserverConfig.AddrManager = test.mockAddrManager
+			}
+			if test.mockSyncManager != nil {
+				rpcserverConfig.SyncMgr = test.mockSyncManager
+			}
+			if test.mockConnManager != nil {
+				rpcserverConfig.ConnMgr = test.mockConnManager
+			}
+			if test.mockClock != nil {
+				rpcserverConfig.Clock = test.mockClock
+			}
 
-		testServer := &rpcServer{cfg: rpcserverConfig}
-		result, err := test.handler(nil, testServer, test.cmd)
-		if test.wantErr {
-			var rpcErr *dcrjson.RPCError
-			if !errors.As(err, &rpcErr) || rpcErr.Code != test.errCode {
-				t.Errorf("%s\nwant: %+v\n got: %+v\n", test.name, test.errCode, rpcErr.Code)
+			testServer := &rpcServer{cfg: *rpcserverConfig}
+			result, err := test.handler(nil, testServer, test.cmd)
+			if test.wantErr {
+				var rpcErr *dcrjson.RPCError
+				if !errors.As(err, &rpcErr) || rpcErr.Code != test.errCode {
+					t.Errorf("%s\nwant: %+v\n got: %+v\n", test.name, test.errCode, rpcErr.Code)
+				}
```

Comment:
I just noticed this as writing a test, but if the error fails as a result of the first condition, `rpcErr.Code` will be nil and cause a panic.

Reply:
Is this the case where `test.wantErr` is `true` but `err` is `nil`? I could modify that section to this to handle if that is the case that you are referring to:
```
if rpcErr != nil {
  t.Errorf("%s\nwant: %+v\n got: %+v\n", test.name, test.errCode, rpcErr.Code)
} else {
  t.Errorf("%s\nwant: %+v\n got: nil\n", test.name, test.errCode)
}
```

Reply:
Yes, either nil or not `*dcrjson.RPCError`. You could also just print out the entire actual err instead of just the code. Either way it would be more helpful to get the error log than a panic... maybe.

Reply:
Added a separate commit for this (so that it doesn't require a re-review of the prior commits, and since it's not really related to the other changes anyway): https://github.com/decred/dcrd/pull/2257/commits/fc5032deb7b210c8ead2cb7fdbef3b9b6746ec82